File: frontend/apps/android/app/src/main/java/app/pantopus/android/ui/screens/shared/list_of_rows/RowModel.kt
@file:Suppress("PackageNaming", "LongParameterList")

package app.pantopus.android.ui.screens.shared.list_of_rows

import androidx.compose.ui.graphics.Color
import app.pantopus.android.ui.components.IdentityPillar
import app.pantopus.android.ui.components.StatusChipVariant
import app.pantopus.android.ui.theme.PantopusIcon

// MARK: - Templates

/** Visual template for a list row. */
enum class RowTemplate { StatusChip, FileChevron, AvatarKebab }

// MARK: - Supporting types (T5 additions)

/** Two-stop gradient for category / avatar / thumbnail backgrounds. */
data class GradientPair(
    val start: Color,
    val end: Color,
)

/** Size variants for [RowLeading.AvatarWithBadge]. */
enum class AvatarBadgeSize(val sizeDp: Int) {
    /** 36dp — inside grouped sections (Discover hub). */
    Small(36),

    /** 40dp — Review claims rows. */
    Medium(40),

    /** 44dp — Connections row. */
    Large(44),
}

/** Background fill for [RowLeading.AvatarWithBadge]. */
sealed interface AvatarBackground {
    data class Solid(val color: Color) : AvatarBackground

    data class Gradient(val gradient: GradientPair) : AvatarBackground
}

/** Thumbnail payload for [RowLeading.Thumbnail]. */
sealed interface ThumbnailImage {
    data class IconOnGradient(
        val icon: PantopusIcon,
        val gradient: GradientPair,
    ) : ThumbnailImage

    data class Remote(
        val url: String,
        val fallback: PantopusIcon,
        val gradient: GradientPair,
    ) : ThumbnailImage
}

/** Size variants for [RowLeading.Thumbnail]. */
enum class ThumbnailSize(val sizeDp: Int) {
    /** 56dp — Offers row. */
    Medium(56),

    /** 64dp — Pets row. */
    Large(64),
}

/** Tone palette for [Bidder] avatars inside [RowLeading.BidderStack]. */
enum class BidderTone { Sky, Teal, Amber, Rose, Violet, Slate }

/** Mini-avatar inside [BidderStack][RowLeading.BidderStack]. */
data class Bidder(
    val id: String,
    val initials: String,
    val tone: BidderTone,
)

/**
 * Bidder-stack payload rendered inline on the chip row — used by My
 * tasks V2 (T5.3.2). Separate from [RowLeading.BidderStack] because
 * the design positions the stack next to the status chip rather than
 * in the leading slot (which holds the 40dp category icon).
 */
data class BidderStackData(
    val bidders: List<Bidder>,
    val overflow: Int = 0,
)

// MARK: - Leading

/** Optional leading visual. */
sealed interface RowLeading {
    data object None : RowLeading

    /** Existing — icon-only. */
    data class Icon(
        val icon: PantopusIcon,
        val tint: Color,
    ) : RowLeading

    /** Existing — avatar with identity-pillar ring. */
    data class Avatar(
        val name: String,
        val imageUrl: String?,
        val identity: IdentityPillar,
        val ringProgress: Float,
    ) : RowLeading

    // ─── T5 additions ──────────────────────────────────────────

    /**
     * 40dp rounded-square tile with a tinted background + foreground icon.
     * Used by Notifications (per type) and Bills (receipt icon).
     */
    data class TypeIcon(
        val icon: PantopusIcon,
        val background: Color,
        val foreground: Color,
    ) : RowLeading

    /**
     * 40dp rounded-square tile with a two-stop gradient background + white
     * foreground icon. Used by My bids / My tasks category icons.
     */
    data class CategoryGradientIcon(
        val icon: PantopusIcon,
        val gradient: GradientPair,
    ) : RowLeading

    /**
     * Plain circular avatar with optional 16dp verified-check overlay at
     * the bottom-right. Used by Connections (large) and Review claims
     * (medium) and Discover hub people rows (small).
     */
    data class AvatarWithBadge(
        val name: String,
        val imageUrl: String?,
        val background: AvatarBackground,
        val size: AvatarBadgeSize,
        val verified: Boolean = false,
    ) : RowLeading

    /** Rounded thumbnail (56dp or 64dp) used by Offers and Pets. */
    data class Thumbnail(
        val image: ThumbnailImage,
        val size: ThumbnailSize,
    ) : RowLeading

    /**
     * Overlapping mini-avatars + `+N` overflow tile, used by My tasks
     * bidder summary.
     */
    data class BidderStack(
        val bidders: List<Bidder>,
        val overflow: Int = 0,
    ) : RowLeading
}

// MARK: - Compact button shared variant

/** Variant for compact in-row buttons. Used by [RowTrailing.VerticalActions]
 *  and [RowFooterAction]. */
enum class CompactButtonVariant { Primary, Ghost, Destructive }

/** Single action description for [RowTrailing.VerticalActions]. */
data class VerticalAction(
    val label: String,
    val variant: CompactButtonVariant,
    val onClick: () -> Unit,
)

// MARK: - Trailing

/** Trailing payload — rendered according to the chosen [RowTemplate]. */
sealed interface RowTrailing {
    data object None : RowTrailing

    data object Chevron : RowTrailing

    data object Kebab : RowTrailing

    data class Status(
        val text: String,
        val variant: StatusChipVariant,
    ) : RowTrailing

    // ─── T5 additions ──────────────────────────────────────────

    /** Amount on top + status chip stacked below — Bills. */
    data class AmountWithChip(
        val amount: String,
        val chipText: String,
        val chipVariant: StatusChipVariant,
        val chipIcon: PantopusIcon? = null,
    ) : RowTrailing

    /** Single circular icon-button — Connections "message" CTA. */
    data class CircularAction(
        val icon: PantopusIcon,
        val accessibilityLabel: String,
        val background: Color,
        val foreground: Color,
        val onClick: () -> Unit,
    ) : RowTrailing

    /** Stacked Accept / Ignore pair — Connections pending requests. */
    data class VerticalActions(
        val primary: VerticalAction,
        val secondary: VerticalAction,
    ) : RowTrailing

    /** Price stack — amount on top, optional sublabel below. */
    data class PriceStack(
        val amount: String,
        val sublabel: String? = null,
    ) : RowTrailing
}

// MARK: - Chip

/** A chip rendered inline with the title (Pets species pill) or in the
 *  chip row beneath the body (My posts intent chip, status chip, counter). */
data class RowChip(
    val text: String,
    val icon: PantopusIcon? = null,
    val tint: Tint,
) {
    sealed interface Tint {
        data class Status(val variant: StatusChipVariant) : Tint

        data class Custom(val background: Color, val foreground: Color) : Tint
    }
}

// MARK: - Footer

/** One action inside [RowFooter]. */
data class RowFooterAction(
    val title: String,
    val icon: PantopusIcon? = null,
    val variant: CompactButtonVariant = CompactButtonVariant.Primary,
    /** Flex weight inside the footer row. Default 1. */
    val flex: Int = 1,
    val onClick: () -> Unit,
)

/** Optional in-card footer for a row — 1–3 inline compact buttons separated
 *  from the card body by a hairline. */
data class RowFooter(
    val actions: List<RowFooterAction>,
)

// MARK: - Highlight

/** Optional visual highlight wrapping the whole card. */
sealed interface RowHighlight {
    /** Notification unread row — primary25 background + personalBg border + dot. */
    data object Unread : RowHighlight

    /** Listing-offer "LEADING" row — amber border + badge. */
    data object Leading : RowHighlight

    /** My-posts archived row — 0.78 opacity. */
    data object Archived : RowHighlight

    /**
     * Terminal / non-actionable row — 0.78 opacity. Used by My bids for
     * rejected / withdrawn / expired / task-cancelled rows so the user
     * can scan them at a glance without confusing them for live bids.
     * Same visual effect as [Archived] but semantically distinct so
     * other screens (Review claims, completed offers) can opt in
     * without overloading the "archived" intent.
     */
    data object Muted : RowHighlight
}

// MARK: - RowModel

/**
 * A single row. ViewModels map their DTOs into a list of these.
 *
 * **Backwards compat:** all T5 fields default to `null` / empty so every
 * existing call site compiles unchanged.
 */
data class RowModel(
    val id: String,
    val title: String,
    val subtitle: String? = null,
    val template: RowTemplate,
    val leading: RowLeading = RowLeading.None,
    val trailing: RowTrailing = RowTrailing.None,
    val onTap: () -> Unit = {},
    val onSecondary: (() -> Unit)? = null,
    // ─── T5 additions ──────────────────────────────────────────
    /** Multiline body rendered below the subtitle (notifications, my posts). */
    val body: String? = null,
    /**
     * Optional small icon prefix for the subtitle line. Used by Connections
     * (map-pin in front of locality).
     */
    val subtitleIcon: PantopusIcon? = null,
    /**
     * Optional small icon prefix for the body line. Used by Connections
     * (per-row interaction-type icon: message-circle / wrench / megaphone /
     * user-plus / sparkles).
     */
    val bodyIcon: PantopusIcon? = null,
    /** Chip rendered inline with the title (Pets species pill). */
    val inlineChip: RowChip? = null,
    /** Chip row beneath the body (intent / status / counter chips). */
    val chips: List<RowChip>? = null,
    /** Small far-right time text on the chip row. */
    val timeMeta: String? = null,
    /** Text appended after the chip row, separated with "·". */
    val metaTail: String? = null,
    /** Italic block quote rendered below the row body (offer notes). */
    val note: String? = null,
    /** Optional visual highlight wrapping the row. */
    val highlight: RowHighlight? = null,
    /** Optional in-card footer with 1–3 compact buttons. */
    val footer: RowFooter? = null,
    /**
     * Optional bidder stack rendered inline on the chip line before
     * the [chips]. Used by My tasks V2 — 22dp overlapping avatars
     * with a `+N` overflow tile.
     */
    val bidderStack: BidderStackData? = null,
)

// MARK: - Section

/** Rendering style for a [RowSection]. */
enum class SectionStyle {
    /** Default — matches T1–T4.1 behaviour (rows separated by spacing). */
    Flat,

    /**
     * Rows grouped inside a single rounded card with hairline dividers
     * between them. Used by Discover hub's typed sections.
     */
    Card,
}

/** Optional grouping for the list body. */
data class RowSection(
    val id: String,
    val header: String? = null,
    val rows: List<RowModel>,
    // ─── T5 additions ──────────────────────────────────────────
    val count: Int? = null,
    val onSeeAll: (() -> Unit)? = null,
    val style: SectionStyle = SectionStyle.Flat,
)

File: frontend/apps/ios/Pantopus/Features/Shared/ListOfRows/RowModel.swift
//
//  RowModel.swift
//  Pantopus
//
//  Template-agnostic row payload. Every list-of-rows screen maps its
//  backend DTOs into one of these.
//
//  T5.0 — extended additively for the My posts / My bids / My tasks /
//  Connections / Notifications / Discover hub / Bills / Pets / Offers /
//  Listing offers / Review claims designs. Every existing call site
//  (Notifications v1, MyHomes, MyClaims, Mailbox list/drawers) compiles
//  unchanged because:
//    - existing enum cases are untouched (`.icon`, `.avatar`, `.none` on
//      `RowLeading`; `.statusChip`, `.chevron`, `.kebab`, `.none` on
//      `RowTrailing`; `.statusChip`, `.fileChevron`, `.avatarKebab` on
//      `RowTemplate`),
//    - new `RowModel` fields are optional with `nil` defaults,
//    - `RowSection` retains its original initializer signature and adds
//      new optional `count` / `onSeeAll` / `style` parameters with
//      defaults.
//

import SwiftUI

// swiftlint:disable enum_case_associated_values_count

/// Visual template for a list row.
public enum RowTemplate: Sendable {
    /// Title + optional subtitle + trailing status chip.
    case statusChip
    /// Leading icon + title + trailing chevron — for drill-down lists.
    case fileChevron
    /// Leading avatar + title + trailing kebab (more-horizontal) menu.
    case avatarKebab
}

// MARK: - Leading

/// Two-stop gradient for category / avatar / thumbnail backgrounds.
public struct GradientPair: Sendable, Hashable {
    public let start: Color
    public let end: Color
    public init(start: Color, end: Color) {
        self.start = start
        self.end = end
    }
}

/// Size variants for the avatar-with-verified-badge leading.
public enum AvatarBadgeSize: Sendable, Hashable {
    /// 36pt — used inside grouped sections (Discover hub).
    case small
    /// 40pt — review-claims rows.
    case medium
    /// 44pt — Connections row.
    case large

    public var size: CGFloat {
        switch self {
        case .small: 36
        case .medium: 40
        case .large: 44
        }
    }
}

/// Background fill for `RowLeading.avatarWithBadge`. Either a solid colour
/// or a two-stop gradient — both come from category / identity tokens; no
/// hex literals at call sites.
public enum AvatarBackground: Sendable, Hashable {
    case solid(Color)
    case gradient(GradientPair)
}

/// Thumbnail payload for `RowLeading.thumbnail` — Offers (56pt) and Pets
/// (64pt) both use this shape.
public enum ThumbnailImage: Sendable, Hashable {
    /// Icon over a gradient fill — fallback when no photo is available.
    case icon(PantopusIcon, gradient: GradientPair)
    /// Remote photo; icon-over-gradient renders while it loads or on
    /// failure.
    case url(URL, fallback: PantopusIcon, gradient: GradientPair)
}

/// Size variants for `RowLeading.thumbnail`.
public enum ThumbnailSize: Sendable, Hashable {
    /// 56pt — Offers row.
    case medium
    /// 64pt — Pets row.
    case large

    public var size: CGFloat {
        switch self {
        case .medium: 56
        case .large: 64
        }
    }
}

/// Mini-avatar used inside the bidder stack on My tasks rows. Tone drives
/// background + foreground colour from the design-token palette.
public struct Bidder: Sendable, Hashable, Identifiable {
    public let id: String
    public let initials: String
    public let tone: BidderTone

    public init(id: String, initials: String, tone: BidderTone) {
        self.id = id
        self.initials = initials
        self.tone = tone
    }
}

/// Categorical tones for `Bidder` avatars. The shell maps each tone to a
/// (background, foreground) colour pair — see `BidderStack.swift`.
public enum BidderTone: Sendable, Hashable, CaseIterable {
    case sky, teal, amber, rose, violet, slate
}

/// Optional leading visual.
public enum RowLeading: Sendable {
    /// Existing — icon-only.
    case icon(PantopusIcon, tint: Color = Theme.Color.primary600)
    /// Existing — avatar with identity-pillar ring.
    case avatar(name: String, imageURL: URL?, identity: IdentityPillar, ringProgress: Double)
    /// Existing — render nothing.
    case none

    // MARK: T5 additions

    /// 40pt rounded-square tile with a tinted background + foreground.
    /// Used by Notifications (per type) and Bills (receipt icon).
    case typeIcon(PantopusIcon, background: Color, foreground: Color)

    /// 40pt rounded-square tile with a two-stop gradient background +
    /// white foreground icon. Used by My bids / My tasks category icons.
    case categoryGradientIcon(PantopusIcon, gradient: GradientPair)

    /// Plain circular avatar with optional 16pt verified-check overlay at
    /// the bottom-right. Used by Connections (large) and Review claims
    /// (medium) and Discover hub people rows (small).
    case avatarWithBadge(
        name: String,
        imageURL: URL?,
        background: AvatarBackground,
        size: AvatarBadgeSize,
        verified: Bool = false
    )

    /// Rounded thumbnail (56pt or 64pt) used by Offers and Pets.
    case thumbnail(image: ThumbnailImage, size: ThumbnailSize)

    /// Overlapping mini-avatars + `+N` overflow tile, used by My tasks
    /// bidder summary.
    case bidderStack(bidders: [Bidder], overflow: Int)
}

// MARK: - Trailing

/// Compact-button variant — used by `RowTrailing.verticalActions` and
/// `RowFooterAction`. Same palette as `CompactButton`.
public enum CompactButtonVariant: Sendable, Hashable {
    case primary
    case ghost
    case destructive
}

/// Single action description for `RowTrailing.verticalActions` (Connections
/// pending-request Accept / Ignore stack).
public struct VerticalAction: Sendable {
    public let label: String
    public let variant: CompactButtonVariant
    public let handler: @Sendable () -> Void

    public init(
        label: String,
        variant: CompactButtonVariant,
        handler: @escaping @Sendable () -> Void
    ) {
        self.label = label
        self.variant = variant
        self.handler = handler
    }
}

/// Trailing payload — rendered according to the chosen `RowTemplate`.
public enum RowTrailing: Sendable {
    /// Existing — status chip.
    case statusChip(text: String, variant: StatusChipVariant)
    /// Existing — drill-down chevron.
    case chevron
    /// Existing — more-horizontal kebab menu.
    case kebab
    /// Existing — render nothing.
    case none

    // MARK: T5 additions

    /// Amount on top + status chip stacked below — Bills.
    case amountWithChip(
        amount: String,
        chipText: String,
        chipVariant: StatusChipVariant,
        chipIcon: PantopusIcon? = nil
    )

    /// Single circular icon-button — Connections "message" CTA.
    case circularAction(
        icon: PantopusIcon,
        accessibilityLabel: String,
        background: Color = Theme.Color.primary50,
        foreground: Color = Theme.Color.primary600,
        handler: @Sendable () -> Void
    )

    /// Stacked Accept / Ignore pair (28-30pt compact buttons) —
    /// Connections pending requests.
    case verticalActions(primary: VerticalAction, secondary: VerticalAction)

    /// Price stack — amount on top, optional sublabel below. Used by My
    /// bids (bid amount + budget). The bid-card title sits to the left.
    case priceStack(amount: String, sublabel: String? = nil)
}

// MARK: - Chip

/// A chip rendered inline with the title (e.g. Pets "Dog" pill) or in the
/// chip row beneath the body (e.g. My posts intent chip, Offers
/// counter-pill).
public struct RowChip: Sendable, Hashable {
    public enum Tint: Sendable, Hashable {
        case status(StatusChipVariant)
        case custom(background: Color, foreground: Color)
    }

    public let text: String
    public let icon: PantopusIcon?
    public let tint: Tint

    public init(text: String, icon: PantopusIcon? = nil, tint: Tint) {
        self.text = text
        self.icon = icon
        self.tint = tint
    }
}

// MARK: - Footer

/// One action inside `RowFooter`. The shell renders all actions as
/// `CompactButton.footer` (34pt) in a horizontal stack, distributed by
/// `flex` weight.
public struct RowFooterAction: Sendable {
    public let title: String
    public let icon: PantopusIcon?
    public let variant: CompactButtonVariant
    /// Flex weight inside the footer row. Default 1.
    public let flex: Int
    public let handler: @Sendable () -> Void

    public init(
        title: String,
        icon: PantopusIcon? = nil,
        variant: CompactButtonVariant = .primary,
        flex: Int = 1,
        handler: @escaping @Sendable () -> Void
    ) {
        self.title = title
        self.icon = icon
        self.variant = variant
        self.flex = flex
        self.handler = handler
    }
}

/// Optional in-card footer for a row. Renders 1–3 inline compact buttons
/// separated by a hairline above. Used by My bids, My tasks, Offers,
/// Listing offers, Review claims.
public struct RowFooter: Sendable {
    public let actions: [RowFooterAction]

    public init(actions: [RowFooterAction]) {
        self.actions = actions
    }
}

// MARK: - Highlight

/// Optional visual highlight wrapping the whole card. Layered with the
/// normal row chrome — does not change geometry.
public enum RowHighlight: Sendable, Hashable {
    /// Notification unread row: `primary25` background, `personalBg` border,
    /// 8pt primary dot beside the title.
    case unread
    /// Listing-offer "LEADING" row: amber border + `LEADING` badge pinned
    /// to the top of the card.
    case leading
    /// My-posts archived row: 0.78 opacity.
    case archived
}

// MARK: - RowModel

/// A single row. Call sites construct these from DTOs in their ViewModel.
///
/// **Backwards compat:** the original 8-parameter init is the
/// `convenience` shape. All new T5 fields default to `nil` and the
/// memberwise init is replaced with a labeled one whose parameter order
/// preserves the original prefix.
public struct RowModel: Identifiable, Sendable {
    public let id: String
    public let title: String
    public let subtitle: String?
    public let template: RowTemplate
    public let leading: RowLeading
    public let trailing: RowTrailing
    /// Invoked when the row is tapped.
    public let onTap: @Sendable () -> Void
    /// Invoked when the kebab menu is tapped (if any).
    public let onSecondary: (@Sendable () -> Void)?

    // MARK: T5 additions

    /// Multiline body rendered below the subtitle. Used by Notifications
    /// (2-line clipped) and My posts (2-line clipped).
    public let body: String?

    /// Optional small icon prefix for the subtitle line. Used by
    /// Connections (map-pin in front of locality).
    public let subtitleIcon: PantopusIcon?

    /// Optional small icon prefix for the body line. Used by Connections
    /// (per-row interaction-type icon: message-circle / wrench / megaphone /
    /// user-plus / sparkles).
    public let bodyIcon: PantopusIcon?

    /// Chip rendered inline with the title (Pets "Dog" pill).
    public let inlineChip: RowChip?

    /// Chip row below the body (My posts intent chip; My bids / My tasks
    /// status chip; Offers counter pill). Renders left-to-right.
    public let chips: [RowChip]?

    /// Small dim text on the far-right of the chip row (My posts "2h",
    /// Notifications "12m").
    public let timeMeta: String?

    /// Text appended after the chip row, separated from chips with a "·"
    /// (My bids "3 others bid · 1d left").
    public let metaTail: String?

    /// Italic block quote rendered below the row body (Listing offers
    /// buyer note).
    public let note: String?

    /// Optional visual highlight wrapping the row (unread / leading /
    /// archived).
    public let highlight: RowHighlight?

    /// Optional in-card footer with 1–3 compact buttons.
    public let footer: RowFooter?

    public init(
        id: String,
        title: String,
        subtitle: String? = nil,
        template: RowTemplate,
        leading: RowLeading = .none,
        trailing: RowTrailing = .none,
        onTap: @escaping @Sendable () -> Void = {},
        onSecondary: (@Sendable () -> Void)? = nil,
        body: String? = nil,
        subtitleIcon: PantopusIcon? = nil,
        bodyIcon: PantopusIcon? = nil,
        inlineChip: RowChip? = nil,
        chips: [RowChip]? = nil,
        timeMeta: String? = nil,
        metaTail: String? = nil,
        note: String? = nil,
        highlight: RowHighlight? = nil,
        footer: RowFooter? = nil
    ) {
        self.id = id
        self.title = title
        self.subtitle = subtitle
        self.template = template
        self.leading = leading
        self.trailing = trailing
        self.onTap = onTap
        self.onSecondary = onSecondary
        self.body = body
        self.subtitleIcon = subtitleIcon
        self.bodyIcon = bodyIcon
        self.inlineChip = inlineChip
        self.chips = chips
        self.timeMeta = timeMeta
        self.metaTail = metaTail
        self.note = note
        self.highlight = highlight
        self.footer = footer
    }
}

// MARK: - Section

/// Style for a `RowSection`. `flat` is the default and matches T1–T4.1
/// behaviour (rows with surface backgrounds, spaced by `s2`); `card`
/// groups all rows inside a single rounded card with hairline separators
/// between them — used by Discover hub's typed sections (People /
/// Businesses / Gigs / Listings).
public enum SectionStyle: Sendable {
    case flat
    case card
}

/// Optional grouping for the list body.
public struct RowSection: Identifiable, Sendable {
    public let id: String
    public let header: String?
    public let rows: [RowModel]

    // MARK: T5 additions

    /// Counter displayed after the section header (Discover hub
    /// "People (24)").
    public let count: Int?
    /// Trailing CTA on the section header (Discover hub "See all →").
    public let onSeeAll: (@Sendable () -> Void)?
    /// Rendering style — default `.flat` preserves T1–T4.1 behaviour.
    public let style: SectionStyle

    public init(
        id: String = UUID().uuidString,
        header: String? = nil,
        rows: [RowModel],
        count: Int? = nil,
        onSeeAll: (@Sendable () -> Void)? = nil,
        style: SectionStyle = .flat
    ) {
        self.id = id
        self.header = header
        self.rows = rows
        self.count = count
        self.onSeeAll = onSeeAll
        self.style = style
    }
}
